In [1]:
# Using Simulacrum V2 instead
import os
import pandas as pd
import warnings 

## 1. Read the datasets with utf-8 encoding and make all column names lower case

In [76]:
def read_file_with_encoding(file_path):
    encodings = ['utf-8', 'latin1', 'ISO-8859-1', 'cp1252']
    
    for encoding in encodings:
        try:
            df = pd.read_csv(file_path, encoding=encoding)
            return df
        except UnicodeDecodeError:
            continue
    
    raise ValueError(f"Failed to decode {file_path} with available encodings.")

In [77]:
def read_file_utf8(file_path):
    try:
        with warnings.catch_warnings(record=True) as w:
            warnings.simplefilter("always", pd.errors.DtypeWarning)
            df = pd.read_csv(file_path, encoding='utf-8', low_memory=False)
            
            # Check if any warnings were triggered
            if any(issubclass(warn.category, pd.errors.DtypeWarning) for warn in w):
                print(f"DtypeWarning encountered in file: {file_path}")
            
        return df
    
    except UnicodeDecodeError:
        raise ValueError(f"Failed to decode {file_path} with UTF-8 encoding.")

In [78]:
directory = "../datasets/simulacrum_v2.1.0/Data"
# directory = "/Users/janexu/Documents/0. 2024 3rd Year/SCIE30002/simulacrum_v2.1.0/Data"
files = os.listdir(directory)

for file_name in files:
    file_path = os.path.join(directory, file_name)
    
    try:
        # Read the file with encoding handling
        df = read_file_with_encoding(file_path)
        
        # Convert column names to lowercase
        df.columns = [col.lower() for col in df.columns]
        
        # Save the updated file
        df.to_csv(file_path, index=False, encoding = 'utf-8')
        
        print(f"Successfully processed and saved: {file_name}")
        
    except Exception as e:
        print(f"Error processing {file_name}: {e}")

/var/folders/wm/5g809ssj2tbbh5kq8m3h9v5m0000gn/T/ipykernel_87148/3401075425.py:6: DtypeWarning: Columns (1,6) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(file_path, encoding=encoding)


Successfully processed and saved: sim_rtds_exposure.csv


/var/folders/wm/5g809ssj2tbbh5kq8m3h9v5m0000gn/T/ipykernel_87148/3401075425.py:6: DtypeWarning: Columns (11) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(file_path, encoding=encoding)


Successfully processed and saved: sim_rtds_prescription.csv
Successfully processed and saved: .DS_Store
Successfully processed and saved: sim_av_patient.csv
Successfully processed and saved: sim_sact_outcome.csv
Successfully processed and saved: sim_sact_drug_detail.csv


/var/folders/wm/5g809ssj2tbbh5kq8m3h9v5m0000gn/T/ipykernel_87148/3401075425.py:6: DtypeWarning: Columns (14,15,16,17,23,24,25,26) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(file_path, encoding=encoding)


Successfully processed and saved: sim_av_tumour.csv
Successfully processed and saved: sim_sact_regimen_new.csv
Successfully processed and saved: sim_av_gene.csv


/var/folders/wm/5g809ssj2tbbh5kq8m3h9v5m0000gn/T/ipykernel_87148/3401075425.py:6: DtypeWarning: Columns (2) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(file_path, encoding=encoding)


Successfully processed and saved: sim_rtds_episode.csv


/var/folders/wm/5g809ssj2tbbh5kq8m3h9v5m0000gn/T/ipykernel_87148/3401075425.py:6: DtypeWarning: Columns (5) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(file_path, encoding=encoding)


Successfully processed and saved: sim_sact_cycle.csv
Successfully processed and saved: sim_sact_regimen.csv


/var/folders/wm/5g809ssj2tbbh5kq8m3h9v5m0000gn/T/ipykernel_87148/3401075425.py:6: DtypeWarning: Columns (15,20) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(file_path, encoding=encoding)


Successfully processed and saved: sim_rtds_combined.csv


## 2. Insert underscores in column names for REDCap upload
SACT Files don't need renaming as they all already have proper underscores

#### AV Files (Patient, Tumour, Gene)

In [79]:
sim_av_patient = pd.read_csv(os.path.join(directory, 'sim_av_patient.csv'))

new_field_names = {
    'patientid': 'patient_id',
    'gender': 'gender_patient',
    'ethnicity': 'ethnicity_patient',
    'deathcausecode_1a': 'death_cause_code_1a',
    'deathcausecode_1b': 'death_cause_code_1b',
    'deathcausecode_1c': 'death_cause_code_1c',
    'deathcausecode_2': 'death_cause_code_2',
    'deathcausecode_underlying': 'death_cause_code_underlying',
    'deathlocationcode': 'death_location_code',
    'vitalstatus': 'vital_status',
    'vitalstatusdate': 'vital_status_date',
    'linknumber': 'link_number'
}

sim_av_patient.rename(columns=new_field_names, inplace=True)
sim_av_patient.to_csv(os.path.join(directory, 'sim_av_patient.csv'), index=False, encoding = 'utf-8')

In [80]:
sim_av_tumour = pd.read_csv(os.path.join(directory, 'sim_av_tumour.csv'))

new_field_names = {
    'tumourid': 'tumour_id',
    'patientid': 'patient_id',
    'gender': 'gender_patient', #unsure about this one
    'diagnosisdatebest': 'diagnosis_date_best',
    'site_icd10r4_o2_3char_from2013': 'site_icd10r4_o2_3char_from_2013',
    'site_icd10r4_o2_from2013': 'site_icd10r4_o2_from_2013',
    'site_icdo3rev2011': 'site_icdo3rev_2011',
    'morph_icdo3rev2011': 'morph_icdo3rev_2011',
    'behaviour_icdo3rev2011': 'behaviour_icdo3rev_2011',
    'screeningstatusfull_code': 'screening_status_full_code',
    'cancercareplanintent': 'cancer_care_plan_intent',
    'performancestatus': 'performance_status'
}

sim_av_tumour.rename(columns=new_field_names, inplace=True)
sim_av_tumour.to_csv(os.path.join(directory, 'sim_av_tumour.csv'), index=False, encoding = 'utf-8')

/var/folders/wm/5g809ssj2tbbh5kq8m3h9v5m0000gn/T/ipykernel_87148/1954489509.py:1: DtypeWarning: Columns (14,15,16,17,23,24,25,26) have mixed types. Specify dtype option on import or set low_memory=False.
  sim_av_tumour = pd.read_csv(os.path.join(directory, 'sim_av_tumour.csv'))


In [81]:
sim_av_gene = pd.read_csv(os.path.join(directory, 'sim_av_gene.csv'))

new_field_names = {
    'geneid': 'gene_id',
    'tumourid': 'tumour_id',
    'patientid': 'patient_id',
    'all_teststatuses': 'all_test_statuses'
}

sim_av_gene.rename(columns=new_field_names, inplace=True)
sim_av_gene.to_csv(os.path.join(directory, 'sim_av_gene.csv'), index=False, encoding = 'utf-8')

#### RTDS Files (Episode, Exposure, Prescription, Combined)

In [82]:
sim_rtds_episode = pd.read_csv(os.path.join(directory, 'sim_rtds_episode.csv'))

new_field_names = {
    'patientid': 'patient_id',
    'radiotherapyepisodeid': 'radiotherapy_episode_id',
    'attendid': 'attend_id',
    'apptdate': 'appt_date',
    'linkcode': 'link_code',
    'decisiontotreatdate': 'decision_to_treat_date',
    'earliestclinappropdate': 'earliest_clin_approp_date',
    'radiotherapypriority': 'radiotherapy_priority',
    'radiotherapyintent': 'radiotherapy_intent'
}

sim_rtds_episode.rename(columns=new_field_names, inplace=True)
sim_rtds_episode.to_csv(os.path.join(directory, 'sim_rtds_episode.csv'), index=False, encoding = 'utf-8')

/var/folders/wm/5g809ssj2tbbh5kq8m3h9v5m0000gn/T/ipykernel_87148/4051879955.py:1: DtypeWarning: Columns (2) have mixed types. Specify dtype option on import or set low_memory=False.
  sim_rtds_episode = pd.read_csv(os.path.join(directory, 'sim_rtds_episode.csv'))


In [83]:
sim_rtds_prescription = pd.read_csv(os.path.join(directory, 'sim_rtds_prescription.csv'))

new_field_names = {
    'patientid': 'patient_id',
    'prescriptionid': 'prescription_id',
    'rttreatmentmodality': 'rt_treatment_modality',
    'rtprescribeddose': 'rt_prescribed_dose',
    'rtprescribedfractions': 'rt_prescribed_fractions',
    'rtactualdose': 'rt_actual_dose',
    'rtactualfractions': 'rt_actual_fractions',
    'rttreatmentregion': 'rt_treatment_region',
    'rttreatmentanatomicalsite': 'rt_treatment_anatomical_site',
    'radiotherapyepisodeid': 'radiotherapy_episode_id',
    'linkcode': 'link_code',
    'attendid': 'attend_id',
    'apptdate': 'appt_date',
}

sim_rtds_prescription.rename(columns=new_field_names, inplace=True)
sim_rtds_prescription.to_csv(os.path.join(directory, 'sim_rtds_prescription.csv'), index=False, encoding = 'utf-8')

/var/folders/wm/5g809ssj2tbbh5kq8m3h9v5m0000gn/T/ipykernel_87148/3852684677.py:1: DtypeWarning: Columns (11) have mixed types. Specify dtype option on import or set low_memory=False.
  sim_rtds_prescription = pd.read_csv(os.path.join(directory, 'sim_rtds_prescription.csv'))


In [84]:
sim_rtds_exposure = pd.read_csv(os.path.join(directory, 'sim_rtds_exposure.csv'))

new_field_names = {
    'prescriptionid': 'prescription_id',
    'radioisotope': 'radio_isotope',
    'rtprescribeddose': 'rt_prescribed_dose',
    'radiotherapybeamtype': 'radiotherapy_beam_type',
    'radiotherapybeamenergy': 'radiotherapy_beam_energy',
    'timeofexposure': 'time_of_exposure',
    'apptdate': 'appt_date',
    'attendid': 'attend_id',
    'patientid': 'patient_id',
    'radiotherapyepisodeid': 'radiotherapy_episode_id',
    'linkcode': 'link_code',
}

sim_rtds_exposure.rename(columns=new_field_names, inplace=True)
sim_rtds_exposure.to_csv(os.path.join(directory, 'sim_rtds_exposure.csv'), index=False, encoding = 'utf-8')

/var/folders/wm/5g809ssj2tbbh5kq8m3h9v5m0000gn/T/ipykernel_87148/3258467019.py:1: DtypeWarning: Columns (1,6) have mixed types. Specify dtype option on import or set low_memory=False.
  sim_rtds_exposure = pd.read_csv(os.path.join(directory, 'sim_rtds_exposure.csv'))


In [85]:
sim_rtds_combined = pd.read_csv(os.path.join(directory, 'sim_rtds_combined.csv'))
# a df combining prescription, episode and exposure (contains all their columns)

new_field_names = {
    'patientid': 'patient_id',
    'prescriptionid': 'prescription_id', # from prescription
    'rttreatmentmodality': 'rt_treatment_modality', # prescription
    'radiotherapypriority': 'radiotherapy_priority', # episode
    'radiotherapyintent': 'radiotherapy_intent', # episode
    'rtprescribeddose': 'rt_prescribed_dose', # prescription
    'rtprescribedfractions': 'rt_prescribed_fractions', # prescription
    'rtactualdose': 'rt_actual_dose', # prescription
    'rtactualfractions': 'rt_actual_fractions', # prescription
    'rttreatmentregion': 'rt_treatment_region', # prescription
    'rttreatmentanatomicalsite': 'rt_treatment_anatomical_site', # prescription
    'decisiontotreatdate': 'decision_to_treat_date', # episode
    'earliestclinappropdate': 'earliest_clin_approp_date', # episode
    'radiotherapyepisodeid': 'radiotherapy_episode_id', # episode
    'linkcode': 'link_code',
    'radioisotope': 'radio_isotope', # exposure
    'radiotherapybeamtype': 'radiotherapy_beam_type', # exposure
    'radiotherapybeamenergy': 'radiotherapy_beam_energy', # exposure
    'timeofexposure': 'time_of_exposure', # exposure
    'apptdate': 'appt_date',
    'attendid': 'attend_id',
    'patientid': 'patient_id',
}

sim_rtds_combined.rename(columns=new_field_names, inplace=True)
sim_rtds_combined.to_csv(os.path.join(directory, 'sim_rtds_combined.csv'), index=False, encoding = 'utf-8')

/var/folders/wm/5g809ssj2tbbh5kq8m3h9v5m0000gn/T/ipykernel_87148/2751768707.py:1: DtypeWarning: Columns (15,20) have mixed types. Specify dtype option on import or set low_memory=False.
  sim_rtds_combined = pd.read_csv(os.path.join(directory, 'sim_rtds_combined.csv'))


## 3. Preprocessing

### Reading in the Newly Modified Files

In [2]:
directory = "../datasets/simulacrum_v2.1.0/Data"
#directory = "/Users/janexu/Documents/0. 2024 3rd Year/SCIE30002/simulacrum_v2.1.0/Data"

# SACT Files
sim_sact_regimen = pd.read_csv(os.path.join(directory, 'sim_sact_regimen.csv')) # shape: (781,389 x 12)
sim_sact_cycle = pd.read_csv(os.path.join(directory, 'sim_sact_cycle.csv')) # shape: (2,741,674 x 6)
sim_sact_drug_detail = pd.read_csv(os.path.join(directory, 'sim_sact_drug_detail.csv')) # shape: (7,662,030 x 7)
sim_sact_outcome = pd.read_csv(os.path.join(directory, 'sim_sact_outcome.csv')) # shape: (784,135 x 6)

# AV Files
sim_av_patient = pd.read_csv(os.path.join(directory, 'sim_av_patient.csv')) # shape: (1,871,605 x 12)
sim_av_tumour = pd.read_csv(os.path.join(directory, 'sim_av_tumour.csv')) # shape: (1,995,570 x 37)
sim_av_gene = pd.read_csv(os.path.join(directory, 'sim_av_gene.csv')) # shape: (255,728 x 29)

# Note: not all tumours have corresponding gene records

# RTDS File
sim_rtds_combined = pd.read_csv(os.path.join(directory, 'sim_rtds_combined.csv'))

/var/folders/wm/5g809ssj2tbbh5kq8m3h9v5m0000gn/T/ipykernel_64702/3344625479.py:6: DtypeWarning: Columns (5) have mixed types. Specify dtype option on import or set low_memory=False.
  sim_sact_cycle = pd.read_csv(os.path.join(directory, 'sim_sact_cycle.csv')) # shape: (2,741,674 x 6)
/var/folders/wm/5g809ssj2tbbh5kq8m3h9v5m0000gn/T/ipykernel_64702/3344625479.py:12: DtypeWarning: Columns (14,15,16,17,23,24,25,26) have mixed types. Specify dtype option on import or set low_memory=False.
  sim_av_tumour = pd.read_csv(os.path.join(directory, 'sim_av_tumour.csv')) # shape: (1,995,570 x 37)
/var/folders/wm/5g809ssj2tbbh5kq8m3h9v5m0000gn/T/ipykernel_64702/3344625479.py:18: DtypeWarning: Columns (15,20) have mixed types. Specify dtype option on import or set low_memory=False.
  sim_rtds_combined = pd.read_csv(os.path.join(directory, 'sim_rtds_combined.csv'))


### Merging (SACT & AV)

In [3]:
# Merging the AV Files (all good)
tumour_patient = pd.merge(sim_av_tumour, sim_av_patient, on='patient_id', how='left') # shape: (1,995,570 x 48)
av = pd.merge(tumour_patient, sim_av_gene, on='tumour_id', how='left') # shape: (2,154,804 x 76)

# Merging the SACT Files
drug_cycle = pd.merge(sim_sact_drug_detail, sim_sact_cycle, on='merged_cycle_id', how='left') # shape: (7,662,030 x 12)
drug_cycle_regimen = pd.merge(drug_cycle, sim_sact_regimen, on='merged_regimen_id', how='left') # shape: (7,662,030 x 23)
sact = pd.merge(drug_cycle_regimen, sim_sact_outcome, on='merged_regimen_id', how='left') # shape: (7,662,030 x 28)

# Ensuring that gender values match across av_patient and av_tumour
matching = av['gender_patient_x'] == av['gender_patient_y']

# All genders match
print(matching.all())

True


In [4]:
# Ensure data type is date time
sim_av_tumour['diagnosis_date_best'] = pd.to_datetime(sim_av_tumour['diagnosis_date_best'])
sim_sact_cycle['start_date_of_cycle'] = pd.to_datetime(sim_sact_cycle['start_date_of_cycle'])
sim_sact_regimen['start_date_of_regimen'] = pd.to_datetime(sim_sact_regimen['start_date_of_regimen'])
sim_rtds_combined['decision_to_treat_date'] = pd.to_datetime(sim_rtds_combined['decision_to_treat_date'])
sim_rtds_combined['appt_date'] = pd.to_datetime(sim_rtds_combined['appt_date'])
sim_rtds_combined['earliest_clin_approp_date'] = pd.to_datetime(sim_rtds_combined['earliest_clin_approp_date'])

pre_2013_tumour = sim_av_tumour[sim_av_tumour['diagnosis_date_best'].dt.year < 2013] # 0 rows
post_2013_tumour = sim_av_tumour[sim_av_tumour['diagnosis_date_best'].dt.year >= 2013] # All rows from the tumour dataset

# As all of the diagnosis dates are above 2013, we will use site_icd10r4_o2_from_2013 to determine the location of the tumour

### Removing Invalid Records

Code for cancers:
1. Male genital cancers: C60-63
2. Female genital cancers: C51-58

Gender Codes:
1. M: 1
2. F: 2
3. Indeterminate: 9

In [5]:
# Find records where females have cancer in male genital areas
prostate_cancer_df = av[av['site_icd10r4_o2_3char_from_2013'].isin(['C60', 'C61', 'C62', 'C63'])] # 190,325 records
invalid_prostate = prostate_cancer_df[prostate_cancer_df['gender_patient_x'] == 2] # 2,737 records

# Find records where males have cancer in female genital areas
gyn_cancer = av[av['site_icd10r4_o2_3char_from_2013'].isin(['C51', 'C52', 'C53', 'C54', 'C55', 'C56', 'C57', 'C58'])]
invalid_gyn = gyn_cancer[gyn_cancer['gender_patient_x'] == 1] # 1,050 records

# Initialise a dictionary to store invalid record id's (tumour, patient, regimen)
bad_records = {}
bad_records['invalid_tumour_ids'] = list(invalid_prostate['tumour_id']) + list(invalid_gyn['tumour_id']) # 3,787 invalid tumour id's

In [6]:
## Finding out regimens that start before tumour diagnosis date and those whose deicison to treat date is earlier than the tumour diagnosis date

# This df has more rows than sim_av_tumour due to the one-to-many relationship between patient and regimen
tumour_regimen = pd.merge(sim_av_tumour, sim_sact_regimen, left_on='patient_id', right_on='encore_patient_id', how='left') # shape: (2,501,838, 49)

# 42,273 invalid regimens
regimen_before_treatment = tumour_regimen[
    (tumour_regimen['start_date_of_regimen'] < tumour_regimen['diagnosis_date_best']) &
    tumour_regimen['start_date_of_regimen'].notna() &
    tumour_regimen['diagnosis_date_best'].notna()
]

# 102,110 invalid regimens
decision_before_treatment = tumour_regimen[
    (tumour_regimen['date_decision_to_treat'] < tumour_regimen['diagnosis_date_best']) &
    tumour_regimen['start_date_of_regimen'].notna() &
    tumour_regimen['diagnosis_date_best'].notna()
]

# Put invalid regimen id's in the dictionary
bad_records['invalid_regimen_ids'] = list(regimen_before_treatment['merged_regimen_id']) + list(decision_before_treatment['merged_regimen_id'])

In [101]:
decision_before_treatment[['date_decision_to_treat', 'diagnosis_date_best']]

date_decision_to_treat diagnosis_date_best
1990                2016-02-22          2017-09-16
1993                2017-06-30          2017-09-16
3474                2016-11-16          2018-02-19
3477                2015-03-29          2016-11-29
3624                2011-09-05          2018-10-24
...                        ...                 ...
2501814             2018-10-18          2019-02-05
2501822             2017-08-03          2018-03-19
2501826             2017-10-18          2018-03-19
2501829             2019-03-18          2019-05-27
2501833             2019-04-10          2019-05-27

[102110 rows x 2 columns]

In [94]:
## Finding out patients who died before treatment/diagnosis but still received treatment

# Vital status D4: 'dies before diagnosis', D5: 'dies before a treatment received'
pre_deceased_patients = sim_av_patient[sim_av_patient['vital_status'].isin(['D4', 'D5'])] # shape: (1169 x 12)

# Merge sact regimen and cycle as the cycle file does not have patient ID
regimen_cycle = pd.merge(sim_sact_regimen, sim_sact_cycle, on='merged_regimen_id', how='left') # shape: (2,766,468 x 17)

# Only include regimens that have corresponding cycles (a.k.a treatments)
filtered_regimen_cycle = regimen_cycle[regimen_cycle['merged_cycle_id'].notna()] # shape: (2,741,674 x 17) [lost 24,794]

# Find number of deceased patients who received treatment (287 records)
invalid_patients = pre_deceased_patients[pre_deceased_patients['patient_id'].isin(filtered_regimen_cycle['encore_patient_id'])]

# Get their patient IDs
# invalid_patients

In [95]:
### Attempt 2
## Finding out patients who have activity after death

# Find patients who are deceased (635,048 records)
deceased_patients = sim_av_patient[sim_av_patient['vital_status'] == 'D']

# Find deceased patients who received treatment
filtered_regimen_cycle = filtered_regimen_cycle[filtered_regimen_cycle['encore_patient_id'].isin(deceased_patients['patient_id'])]

# Merge with the patient dataset to add date of death
filtered_regimen_cycle = filtered_regimen_cycle.merge(deceased_patients, left_on='encore_patient_id', right_on='patient_id', how='left')

# Find cycles that were conducted after death (422,012)
cycle_after_death = filtered_regimen_cycle[
    (filtered_regimen_cycle['vital_status_date'] < filtered_regimen_cycle['start_date_of_cycle']) &
    filtered_regimen_cycle['vital_status_date'].notna() &
    filtered_regimen_cycle['start_date_of_cycle'].notna()
]

# Find the number of cycles per regimen
cycle_counts = filtered_regimen_cycle.groupby('merged_regimen_id')['merged_cycle_id'].nunique().reset_index()
cycle_counts.columns = ['merged_regimen_id', 'cycle_count']

# Find regimens that only consist of 1 cycle, and whose cycle was conducted after death
cycle_counts = cycle_counts[cycle_counts['cycle_count'] == 1]
cycle_counts = cycle_counts[cycle_counts['merged_regimen_id'].isin(cycle_after_death['merged_regimen_id'])]

# 29,803 invalid regimens
invalid_regimens = list(cycle_counts['merged_regimen_id'])

# 392,209 cycles remaining
filtered_cycle_after_death = cycle_after_death[~cycle_after_death['merged_regimen_id'].isin(cycle_counts['merged_regimen_id'])]

In [100]:
filtered_cycle_after_death[['merged_cycle_id', 'merged_regimen_id', 'vital_status_date', 'start_date_of_cycle']].head(10)

merged_cycle_id  merged_regimen_id vital_status_date start_date_of_cycle
20       10000030.0           10030686        2016-12-06          2018-11-25
21       10020891.0           10030686        2016-12-06          2019-01-17
22       10035573.0           10030686        2016-12-06          2019-03-21
46       10063278.0           10030717        2020-01-14          2020-01-15
47       10068942.0           10030717        2020-01-14          2020-02-12
48       10073442.0           10030717        2020-01-14          2020-03-11
49       10076989.0           10030717        2020-01-14          2020-04-08
50       10079806.0           10030717        2020-01-14          2020-05-06
51       10082046.0           10030717        2020-01-14          2020-06-03
52       10083832.0           10030717        2020-01-14          2020-07-01

In [96]:
filtered_cycle_after_death.shape

(392209, 29)

In [92]:
filtered.shape

(392209, 29)

(392209, 29)

In [ ]:
## Decision to Treat (sact_regimen) should be later than Tumour Diagnosis Date (av_tumour)

In [68]:
sim_av_patient[sim_av_patient['patient_id'] == 10404139]

patient_id  gender_patient ethnicity_patient death_cause_code_1a  \
404138    10404139               2                 A                J449   

       death_cause_code_1b death_cause_code_1c death_cause_code_2  \
404138                J841                 NaN           G20,F019   

       death_cause_code_underlying death_location_code vital_status  \
404138                        J449                   4            D   

       vital_status_date  link_number  
404138        2016-12-06    101649026

In [69]:
sim_av_tumour[sim_av_tumour['patient_id'] == 10404139]

tumour_id  gender_patient  patient_id diagnosis_date_best  \
404138   10825482               2    10404139          2016-11-16   

       site_icd10_o2_3char site_icd10_o2 site_icd10r4_o2_3char_from_2013  \
404138                 C44          C442                             C44   

       site_icd10r4_o2_from_2013 site_icdo3rev_2011 site_icdo3rev2011_3char  \
404138                      C442               C442                     C44   

        ...  quintile_2019  date_first_surgery  cancer_care_plan_intent  \
404138  ...              3          2016-11-16                        C   

        performance_status chrl_tot_27_03 comorbidities_27_03 gleason_primary  \
404138                 0.0            0.0                 NaN             NaN   

       gleason_secondary gleason_tertiary  gleason_combined  
404138               NaN              NaN               NaN  

[1 rows x 37 columns]

In [66]:
filtered[filtered['merged_regimen_id'] == 10030686]

merged_cycle_id  merged_regimen_id vital_status_date start_date_of_cycle
755       10035573.0           10030686        2016-12-06          2019-03-21

In [63]:
filtered_regimen_cycle[filtered_regimen_cycle['merged_cycle_id'].isin(filtered['merged_cycle_id'])]['merged_regimen_id']

121         10030686
211         10030717
377         10030807
456         10030832
514         10030844
             ...    
2766390    250009340
2766395    250009343
2766435    250009382
2766461    250009404
2766465    250009409
Name: merged_regimen_id, Length: 54702, dtype: int64

In [64]:
filtered_regimen_cycle[filtered_regimen_cycle['merged_regimen_id'] == 10030686]

encore_patient_id  merged_regimen_id  height_at_start_of_regimen  \
119           10404139           10030686                        1.75   
120           10404139           10030686                        1.75   
121           10404139           10030686                        1.75   

     weight_at_start_of_regimen intent_of_treatment date_decision_to_treat  \
119                         NaN                  03             2015-12-31   
120                         NaN                  03             2015-12-31   
121                         NaN                  03             2015-12-31   

    start_date_of_regimen    mapped_regimen clinical_trial chemo_radiation  \
119            2018-11-18  Hydroxycarbamide             99             NaN   
120            2018-11-18  Hydroxycarbamide             99             NaN   
121            2018-11-18  Hydroxycarbamide             99             NaN   

      benchmark_group  link_number  merged_cycle_id  cycle_number  \
119  HYDROXYCARBAMIDE    101649026       10000030.0           1.0   
120  HYDROXYCARBAMIDE    101649026       10020891.0           2.0   
121  HYDROXYCARBAMIDE    101649026       10035573.0          36.0   

    start_date_of_cycle opcs_procurement_code perf_status_start_of_cycle  
119          2018-11-25                   NaN                          1  
120          2019-01-17                   NaN                          0  
121          2019-03-21                   NaN                          1

In [53]:
invalid_cycles = filtered_regimen_cycle[filtered_regimen_cycle['merged_cycle_id'].isin(dates['merged_cycle_id'])]
invalid_cycles

encore_patient_id  merged_regimen_id  height_at_start_of_regimen  \
121               10404139           10030686                        1.75   
211               10404439           10030717                        0.00   
325               10405333           10030776                         NaN   
377               10405762           10030807                        1.51   
456               10406112           10030832                        0.00   
...                    ...                ...                         ...   
2766390          250002107          250009340                        1.66   
2766395          250002294          250009343                        1.70   
2766435          250000823          250009382                        1.71   
2766461          250002014          250009404                        1.83   
2766465          250001131          250009409                        1.56   

         weight_at_start_of_regimen intent_of_treatment  \
121                             NaN                  03   
211                            68.5                   P   
325                             NaN                   C   
377                             NaN                 NaN   
456                             0.0                   P   
...                             ...                 ...   
2766390                        61.3                 NaN   
2766395                        85.6                 NaN   
2766435                        85.6                 NaN   
2766461                        66.4                   P   
2766465                        89.0                   P   

        date_decision_to_treat start_date_of_regimen          mapped_regimen  \
121                 2015-12-31            2018-11-18        Hydroxycarbamide   
211                 2019-08-21            2019-09-11            Enzalutamide   
325                        NaN            2017-11-17                     NaN   
377                        NaN            2016-05-31        Hydroxycarbamide   
456                 2019-07-16            2019-07-18        Hydroxycarbamide   
...                        ...                   ...                     ...   
2766390                    NaN            2019-07-12               NIVOLUMAB   
2766395                    NaN            2018-12-13       GemCarbo Days 1+8   
2766435                    NaN            2020-10-04   Carboplatin AUC 2 - 5   
2766461             2020-08-19            2020-12-04            Atezolizumab   
2766465             2020-10-04            2020-10-17  Cisplatin + Pemetrexed   

        clinical_trial chemo_radiation            benchmark_group  \
121                 99             NaN           HYDROXYCARBAMIDE   
211                 02               N               ENZALUTAMIDE   
325                NaN             NaN           HYDROXYCARBAMIDE   
377                 02             NaN           HYDROXYCARBAMIDE   
456                 02             NaN           HYDROXYCARBAMIDE   
...                ...             ...                        ...   
2766390            NaN             NaN                  NIVOLUMAB   
2766395            NaN             NaN  CARBOPLATIN + GEMCITABINE   
2766435             02               N                NOT MATCHED   
2766461             02             NaN               ATEZOLIZUMAB   
2766465            NaN             NaN               ATEZOLIZUMAB   

         link_number  merged_cycle_id  cycle_number start_date_of_cycle  \
121        101649026       10035573.0          36.0          2019-03-21   
211        100351577       10091528.0          14.0          2021-07-26   
325        101346822       10000064.0           1.0          2017-11-17   
377        100450428       10091598.0           5.0          2019-01-01   
456        100337644       10090113.0          10.0          2020-12-01   
...              ...              ...           ...                 ...   
2766390    100626278      250007255.0           4.0         

In [52]:
sim_sact_regimen['merged_regimen_id'].nunique()

781389

In [51]:
filtered_regimen_cycle['merged_regimen_id'].nunique()

756595

In [33]:
max_cycle_dates[max_cycle_dates.duplicated(subset=['encore_patient_id', 'start_date_of_cycle'])]

encore_patient_id start_date_of_cycle  merged_regimen_id  \
74               10030463          2021-09-23           10070770   
257              10099479          2022-05-06           10054811   
265              10102796          2019-04-01           10054815   
353              10137644          2022-02-26           10042622   
371              10145200          2017-04-21           10063675   
...                   ...                 ...                ...   
355697          250002246                 NaT          250005961   
355698          250002246                 NaT          250005961   
355699          250002246                 NaT          250007785   
355700          250002246                 NaT          250007785   
355795          250002398          2019-01-13          250008327   

        height_at_start_of_regimen  weight_at_start_of_regimen  \
74                            1.64                        83.0   
257                           1.77                        45.8   
265                           1.84                        75.0   
353                           1.58                        68.5   
371                           1.65                         0.0   
...                            ...                         ...   
355697                        1.60                        57.4   
355698                        1.60                        57.4   
355699                        1.73                        67.8   
355700                        1.73                        67.8   
355795                        0.00                        61.9   

       intent_of_treatment date_decision_to_treat start_date_of_regimen  \
74                      03                    NaN            2021-07-20   
257                    NaN                    NaN            2022-04-15   
265                     03             2016-10-08            2019-02-01   
353                    NaN             2022-01-23            2022-01-29   
371                      P             2016-09-10            2017-04-21   
...                    ...                    ...                   ...   
355697                   P                    NaN                   NaT   
355698                   P                    NaN                   NaT   
355699                   P                    NaN                   NaT   
355700                   P                    NaN                   NaT   
355795                 NaN             2018-12-29            2019-01-13   

                          mapped_regimen clinical_trial chemo_radiation  \
74                              Avelumab             02               N   
257           FLUOROURACIL + OXALIPLATIN             02             NaN   
265     Capecitabine + Oxaliplatin 21day             02               N   
353                          Apalutamide             02             NaN   
371                          DARATUMUMAB             02               N   
...                                  ...            ...             ...   
355697                               NaN             02             NaN   
355698                               NaN             02             NaN   
355699          EW Rel rEECur Ifosfamide             01               N   
355700          EW Rel rEECur Ifosfamide             01               N   
355795               AVELUMAB + AXITINIB             02               N   

                             benchmark_group  link_number  merged_cycle_id  \
74                                  AVELUMAB    100517013       10046247.0   
257               FLUOROURACIL + OXALIPLATIN    100222188       10029607.0   
265               CAPECITABINE + OXALIPLATIN    100157101       10060871.0   
353     DOCETAXEL + PERTUZUMAB + TRASTUZUMAB    101386404       10041486.0   
371                              DARATUMUMAB    101485495       10015559.0   
...                                      ...          ...              ...   
355697                IPILIMUMAB + NIVOLUMAB    10

In [ ]:
### Note: should also do the same for radiotherapy as we did for regimens / cycles

## Radiotherapy that started before tumour diagnosis

## Finding patients who died before radiotherapy but still have radiotherapy records

## Finding patients who died but still received radiotherapy after death

In [121]:
deceased_with_regimen

676        False
8303       False
11124      False
14111      False
17293      False
           ...  
1867367    False
1869336    False
1869496    False
1870028    False
1870199    False
Name: patient_id, Length: 1169, dtype: bool

In [117]:
regimen_cycle['merged_cycle_id'].isna().sum()

24794

### Dropping Unused Columns

In [ ]:
# Drop Start Date of Regimen

# Drop site 2011 from AV Tumour

### Playground

- This is all code we played around with but ultimately didn't end up using in the final product

In [ ]:
sact_regimen_ids = set(sact['merged_regimen_id'].unique())
regimen_regimen_ids = set(sim_sact_regimen['merged_regimen_id'].unique())

# Find regimen id's that aren't in the final merged df
missing_regimen_ids = regimen_regimen_ids - sact_regimen_ids

# Filter the sim_sact_regimen df for these missing regimen_id's
missing_regimens = sim_sact_regimen[sim_sact_regimen['merged_regimen_id'].isin(missing_regimen_ids)]

missing_regimens.head()

encore_patient_id  merged_regimen_id  height_at_start_of_regimen  \
72            10405821           10030812                         0.0   
135           10407709           10030955                         0.0   
179           10409421           10031066                         0.0   
188           10409813           10031091                         0.0   
350           10415146           10031491                         0.0   

     weight_at_start_of_regimen intent_of_treatment date_decision_to_treat  \
72                          0.0                   P             2019-09-25   
135                        55.0                 NaN             2019-12-09   
179                         NaN                   P                    NaN   
188                         NaN                 NaN                    NaN   
350                         NaN                   P                    NaN   

    start_date_of_regimen    mapped_regimen clinical_trial chemo_radiation  \
72             2019-10-24  Hydroxycarbamide             02               N   
135            2019-12-09  Hydroxycarbamide            NaN             NaN   
179            2019-12-01               NaN             02               N   
188            2018-12-28               NaN             02             NaN   
350            2021-04-21               NaN             02             NaN   

      benchmark_group  link_number  
72   HYDROXYCARBAMIDE    101836626  
135  HYDROXYCARBAMIDE    100912531  
179      ENZALUTAMIDE    100316506  
188      ENZALUTAMIDE    101470000  
350  HYDROXYCARBAMIDE    100617606

In [20]:
# Dropping start_date in sim_sact_regimen and just keeping sact_cycle as a more accurate starting date
# sim_sact_regimen.drop(['start_date_of_regimen', 'height_at_start_of_regimen', 'weight_at_start_of_regimen'], axis=1, inplace=True)
# sim_sact_regimen.to_csv(os.path.join(directory, 'sim_sact_regimen_new.csv'), index=False)

#### SACT Files

Unique Regimen ID count:
- In sim_sact_outcome: 784,135
- In sim_sact_regimen: 781,389
- In sim_sact_cycle: 756,595 (24,794 less than regimen)
- In final merged df: 751,911 (29,478 less than regimen)

Unique Cycle ID count:
- In sim_sact_cycle: 2,741,674
- In sim_sact_drug_detail: 2,729,567 (12,107 less than cycle)

Note: there are some discrepancies between the number of unique regimen ID's in the original regimen file and the final merged dataframe. This may be because some records don't have corresponding records in other tables (e.g. some regimens do not have corresponding cycles, and some cycles do not have drug details)

In [22]:
# Meging the SACT Files
regimen_cycle = pd.merge(sim_sact_regimen, sim_sact_cycle, on='merged_regimen_id', how='left') # shape: (2,766,468 x 17)
regimen_cycle_drug = pd.merge(regimen_cycle, sim_sact_drug_detail, on='merged_cycle_id', how='left') # shape: (7,698,931 x 23)
regimen_cycle_drug_outcome = pd.merge(regimen_cycle_drug, sim_sact_outcome, on='merged_regimen_id', how='left') # shape: (7,698,931 x 28)

In [ ]:
# Ensure start date is in date/time format
# filtered_regimen_cycle['start_date_of_cycle'] = pd.to_datetime(filtered_regimen_cycle['start_date_of_cycle'], errors='coerce')

In [42]:
## Finding out patients who have activity after death

# Find patients who are deceased (635,048 records)
deceased_patients = sim_av_patient[sim_av_patient['vital_status'] == 'D']

# Find the latest cycle start date for each patient (348,610 records)
max_cycle_dates = filtered_regimen_cycle.groupby('encore_patient_id')['start_date_of_cycle'].max().reset_index()

# Find the deceased patients who received treatment (162,710 records)
deceased_patients = deceased_patients[deceased_patients['patient_id'].isin(max_cycle_dates['encore_patient_id'])]

# Merge the latest cycle dates (355,882 records) --> there are duplicate cycles that end on the same date but are part of different regimens
max_cycle_dates = max_cycle_dates.merge(filtered_regimen_cycle, on=['encore_patient_id', 'start_date_of_cycle'], how='left')

# Only get records for deceased patients (166,419 records)
max_date_per_patient = max_cycle_dates[max_cycle_dates['encore_patient_id'].isin(deceased_patients['patient_id'])]

dates = pd.merge(deceased_patients, max_date_per_patient, left_on='patient_id', right_on='encore_patient_id')[['merged_cycle_id', 'merged_regimen_id', 'vital_status_date', 'start_date_of_cycle']]

# Find records whose patient's death date is earlier than the treatment date (62,490 records)
dates = dates[dates['vital_status_date'] < dates['start_date_of_cycle']]

dates

merged_cycle_id  merged_regimen_id vital_status_date  \
4            10087813.0           10036517        2021-08-09   
12           10090306.0           10036605        2021-07-12   
36           10087285.0           10063655        2020-08-13   
43           10085234.0           10071388        2021-03-13   
49           10078189.0           10036940        2019-12-13   
...                 ...                ...               ...   
166402      250007243.0          250007881        2018-11-24   
166404      250003148.0          250009248        2020-10-09   
166407      250003867.0          250007205        2019-11-06   
166415      250005550.0          250008898        2020-02-18   
166417      250003873.0          250007219        2019-09-25   

       start_date_of_cycle  
4               2022-05-16  
12              2021-09-03  
36              2021-02-10  
43              2021-04-20  
49              2020-05-12  
...                    ...  
166402          2019-07-28  
166404          2020-11-27  
166407          2020-02-06  
166415          2020-05-05  
166417          2019-12-20  

[62490 rows x 4 columns]